core/src/main/java/hudson/slaves/NodeProvisioner.java

```diff
@@ -217,29 +218,40 @@ public void run() {
-                                        e.getCause());
-                                for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
-                                    cl.onFailure(f, e.getCause());
+                                Node node = null;
+                                try {
+                                    node = f.future.get();
+                                    for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
+                                        cl.onComplete(f, node);
+                                    }
+                                } catch (InterruptedException e) {
+                                    throw new AssertionError(e); // since we confirmed that the future is already done
+                                } catch (ExecutionException e) {
+                                    Throwable cause = e.getCause();
+                                    if(!(cause instanceof AbortException)) {
+                                        LOGGER.log(Level.WARNING, "Unexpected exception encountered while "
+                                                + "provisioning agent " + f.displayName, cause);
+                                    }
+                                    for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
+                                        cl.onFailure(f, cause);
+                                    }
+                                    node = null;
                                 }
-                            } catch (IOException e) {
-                                LOGGER.log(Level.WARNING, "Provisioned agent " + f.displayName + " failed to launch", 
-                                        e);
-                                for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
-                                    cl.onFailure(f, e);
+
+                                try {
+                                    if (node != null) {
+                                        jenkins.addNode(node);
+                                        LOGGER.log(Level.INFO,
+                                                "{0} provisioning successfully completed. "
+                                                        + "We have now {1,number,integer} computer(s)",
+                                                new Object[]{f.displayName, jenkins.getComputers().length});
+                                    }
+                                } catch (IOException e) {
+                                    LOGGER.log(Level.WARNING, "Provisioned agent " + f.displayName + " failed to launch",
+                                            e);
+                                    for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
+                                        cl.onFailure(f, e);
```

Comment:
guard against exceptions thrown here... perhaps the best plan would be to move these loops into a `CloudProvisionlinListener.fireOnFailure(f,e)` and `CloudProvisionlinListener.fireOnComplete(f,node)` static helper methods that can ensure exceptions are not propagated.

Also, we have to ask if it makes sense to fire on complete and then failure where the failure is the IOException saving the node to disk... I wonder should we perform some different behaviour for the failure to save to disk